readthedocs/projects/tasks.py

```diff
@@ -715,6 +723,101 @@ def save_build_config(self):
         })
         self.build['config'] = config
 
+    def store_build_artifacts(
+            self,
+            html=False,
+            localmedia=False,
+            search=False,
+            pdf=False,
+            epub=False,
+    ):
+        """
+        Save build artifacts to "storage" using Django's storage API
+
+        The storage could be local filesystem storage OR cloud blob storage
+        such as S3, Azure storage or Google Cloud Storage.
+
+        Remove build artifacts of types not included in this build (PDF, ePub, zip only).
+
+        This looks very similar to `move_files` and is intended to replace it!
+
+        :param html: whether to save HTML output
+        :param localmedia: whether to save localmedia (htmlzip) output
+        :param search: whether to save search artifacts
+        :param pdf: whether to save PDF output
+        :param epub: whether to save ePub output
+        """
+        if settings.RTD_BUILD_MEDIA_STORAGE:
+            log.info(
+                LOG_TEMPLATE.format(
+                    project=self.version.project.slug,
+                    version=self.version.slug,
+                    msg='Writing build artifacts to media storage',
+                ),
+            )
+
+            storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
+
+            types_to_copy = []
+            types_to_delete = []
+
+            # HTML media
+            if html:
+                types_to_copy.append(('html', self.config.doctype))
+
+            # Search media (JSON)
+            if search:
+                types_to_copy.append(('json', 'sphinx_search'))
+
+            if localmedia:
+                types_to_copy.append(('htmlzip', 'sphinx_localmedia'))
+            else:
+                types_to_delete.append('htmlzip')
+
+            if pdf:
+                types_to_copy.append(('pdf', 'sphinx_pdf'))
+            else:
+                types_to_delete.append('pdf')
+
+            if epub:
+                types_to_copy.append(('epub', 'sphinx_epub'))
+            else:
+                types_to_delete.append('epub')
+
+            for media_type, build_type in types_to_copy:
+                from_path = self.version.project.artifact_path(
+                    version=self.version.slug,
+                    type_=build_type,
+                )
+                to_path = self.version.project.get_storage_path(
+                    type_=media_type,
+                    version_slug=self.version.slug,
+                    include_file=False,
+                )
+                log.info(
+                    LOG_TEMPLATE.format(
+                        project=self.version.project.slug,
+                        version=self.version.slug,
+                        msg=f'Writing {media_type} to media storage - {to_path}',
+                    ),
+                )
+                storage.copy_directory(from_path, to_path)
```

Comment:
This likely should be `try/except` as well, so when it fails it doesn't stop all the other formats from syncing.

Reply:
I added a try block around the writes/deletes to storage just in case.